File: shared/api/fcm-service.go
// FCM Service using Server Key (Go)
package api

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

type FCMService struct {
	ServerKey string
	Client    *http.Client
}

type FCMNotification struct {
	Title string `json:"title"`
	Body  string `json:"body"`
	Sound string `json:"sound,omitempty"`
	Icon  string `json:"icon,omitempty"`
}

type FCMMessage struct {
	To           string          `json:"to,omitempty"`
	Tokens       []string        `json:"registration_ids,omitempty"`
	Topic        string          `json:"to,omitempty"` // For topics, use "/topics/topic_name"
	Notification FCMNotification `json:"notification"`
	Data         map[string]string `json:"data,omitempty"`
}

type FCMResponse struct {
	MulticastID  int64 `json:"multicast_id"`
	Success      int   `json:"success"`
	Failure      int   `json:"failure"`
	CanonicalIDs int   `json:"canonical_ids"`
	Results      []struct {
		MessageID string `json:"message_id,omitempty"`
		Error     string `json:"error,omitempty"`
	} `json:"results,omitempty"`
	MessageID int64  `json:"message_id,omitempty"`
	Error     string `json:"error,omitempty"`
}

// NewFCMService creates a new FCM service instance
func NewFCMService(serverKey string) *FCMService {
	return &FCMService{
		ServerKey: serverKey,
		Client:    &http.Client{},
	}
}

// SendNotification sends a push notification to a single device
func (f *FCMService) SendNotification(deviceToken string, notification FCMNotification, data map[string]string) (*FCMResponse, error) {
	message := FCMMessage{
		To:           deviceToken,
		Notification: notification,
		Data:         data,
	}

	if message.Data == nil {
		message.Data = make(map[string]string)
	}
	message.Data["click_action"] = "FLUTTER_NOTIFICATION_CLICK"

	return f.sendMessage(message)
}

// SendMulticast sends a push notification to multiple devices
func (f *FCMService) SendMulticast(deviceTokens []string, notification FCMNotification, data map[string]string) (*FCMResponse, error) {
	message := FCMMessage{
		Tokens:       deviceTokens,
		Notification: notification,
		Data:         data,
	}

	if message.Data == nil {
		message.Data = make(map[string]string)
	}
	message.Data["click_action"] = "FLUTTER_NOTIFICATION_CLICK"

	return f.sendMessage(message)
}

// SendToTopic sends a push notification to a topic
func (f *FCMService) SendToTopic(topic string, notification FCMNotification, data map[string]string) (*FCMResponse, error) {
	message := FCMMessage{
		Topic:        fmt.Sprintf("/topics/%s", topic),
		Notification: notification,
		Data:         data,
	}

	if message.Data == nil {
		message.Data = make(map[string]string)
	}
	message.Data["click_action"] = "FLUTTER_NOTIFICATION_CLICK"

	return f.sendMessage(message)
}

// sendMessage sends the FCM message
func (f *FCMService) sendMessage(message FCMMessage) (*FCMResponse, error) {
	// Set default values
	if message.Notification.Sound == "" {
		message.Notification.Sound = "default"
	}
	if message.Notification.Icon == "" {
		message.Notification.Icon = "ic_notification"
	}

	// Convert message to JSON
	jsonData, err := json.Marshal(message)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal FCM message: %w", err)
	}

	// Create HTTP request
	req, err := http.NewRequest("POST", "https://fcm.googleapis.com/fcm/send", bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("failed to create HTTP request: %w", err)
	}

	// Set headers
	req.Header.Set("Authorization", fmt.Sprintf("key=%s", f.ServerKey))
	req.Header.Set("Content-Type", "application/json")

	// Send request
	resp, err := f.Client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send FCM request: %w", err)
	}
	defer resp.Body.Close()

	// Read response
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read FCM response: %w", err)
	}

	// Parse response
	var fcmResp FCMResponse
	if err := json.Unmarshal(body, &fcmResp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal FCM response: %w", err)
	}

	// Check for errors
	if resp.StatusCode != http.StatusOK {
		return &fcmResp, fmt.Errorf("FCM request failed with status %d: %s", resp.StatusCode, string(body))
	}

	return &fcmResp, nil
}

// Usage example:
/*
package main

import (
	"fmt"
	"log"
)

func main() {
	// Initialize FCM service
	fcm := NewFCMService("YOUR_FCM_SERVER_KEY_HERE")

	// Send notification to single device
	resp, err := fcm.SendNotification(
		"ePaxeXHZRoGyUnWbMXU2...", // device token
		FCMNotification{
			Title: "Hello from Go-Mailer!",
			Body:  "Your notification message here",
			Sound: "default",
		},
		map[string]string{
			"campaign_id": "welcome_campaign",
			"user_id":     "user123",
		},
	)

	if err != nil {
		log.Printf("❌ Failed to send notification: %v", err)
		return
	}

	if resp.Success > 0 {
		fmt.Printf("✅ Notification sent successfully! Message ID: %v\n", resp.Results[0].MessageID)
	} else {
		fmt.Printf("❌ Failed to send notification: %s\n", resp.Results[0].Error)
	}
}
*/

File: shared/api/fcm-service.js
// FCM Service using Server Key (Node.js)
const https = require('https');

class FCMService {
  constructor(serverKey) {
    this.serverKey = serverKey;
  }

  /**
   * Send push notification to a single device
   * @param {string} deviceToken - FCM device token
   * @param {object} notification - Notification payload
   * @param {object} data - Data payload (optional)
   * @returns {Promise<object>} - FCM response
   */
  async sendNotification(deviceToken, notification, data = {}) {
    const payload = JSON.stringify({
      to: deviceToken,
      notification: {
        title: notification.title,
        body: notification.body,
        sound: notification.sound || 'default',
        icon: notification.icon || 'ic_notification',
      },
      data: {
        ...data,
        click_action: 'FLUTTER_NOTIFICATION_CLICK', // For Flutter
      }
    });

    return new Promise((resolve, reject) => {
      const options = {
        hostname: 'fcm.googleapis.com',
        port: 443,
        path: '/fcm/send',
        method: 'POST',
        headers: {
          'Authorization': `key=${this.serverKey}`,
          'Content-Type': 'application/json',
          'Content-Length': Buffer.byteLength(payload)
        }
      };

      const req = https.request(options, (res) => {
        let data = '';
        
        res.on('data', (chunk) => {
          data += chunk;
        });
        
        res.on('end', () => {
          try {
            const response = JSON.parse(data);
            if (res.statusCode === 200 && response.success === 1) {
              resolve({
                success: true,
                messageId: response.results[0].message_id,
                response: response
              });
            } else {
              reject({
                success: false,
                error: response.results?.[0]?.error || 'Unknown error',
                response: response
              });
            }
          } catch (error) {
            reject({
              success: false,
              error: 'Invalid JSON response',
              rawResponse: data
            });
          }
        });
      });

      req.on('error', (error) => {
        reject({
          success: false,
          error: error.message
        });
      });

      req.write(payload);
      req.end();
    });
  }

  /**
   * Send push notification to multiple devices
   * @param {string[]} deviceTokens - Array of FCM device tokens
   * @param {object} notification - Notification payload
   * @param {object} data - Data payload (optional)
   * @returns {Promise<object>} - FCM response
   */
  async sendMulticast(deviceTokens, notification, data = {}) {
    const payload = JSON.stringify({
      registration_ids: deviceTokens,
      notification: {
        title: notification.title,
        body: notification.body,
        sound: notification.sound || 'default',
        icon: notification.icon || 'ic_notification',
      },
      data: {
        ...data,
        click_action: 'FLUTTER_NOTIFICATION_CLICK',
      }
    });

    return new Promise((resolve, reject) => {
      const options = {
        hostname: 'fcm.googleapis.com',
        port: 443,
        path: '/fcm/send',
        method: 'POST',
        headers: {
          'Authorization': `key=${this.serverKey}`,
          'Content-Type': 'application/json',
          'Content-Length': Buffer.byteLength(payload)
        }
      };

      const req = https.request(options, (res) => {
        let data = '';
        
        res.on('data', (chunk) => {
          data += chunk;
        });
        
        res.on('end', () => {
          try {
            const response = JSON.parse(data);
            resolve({
              success: response.success > 0,
              successCount: response.success,
              failureCount: response.failure,
              response: response
            });
          } catch (error) {
            reject({
              success: false,
              error: 'Invalid JSON response',
              rawResponse: data
            });
          }
        });
      });

      req.on('error', (error) => {
        reject({
          success: false,
          error: error.message
        });
      });

      req.write(payload);
      req.end();
    });
  }

  /**
   * Send notification to a topic
   * @param {string} topic - FCM topic name
   * @param {object} notification - Notification payload
   * @param {object} data - Data payload (optional)
   * @returns {Promise<object>} - FCM response
   */
  async sendToTopic(topic, notification, data = {}) {
    const payload = JSON.stringify({
      to: `/topics/${topic}`,
      notification: {
        title: notification.title,
        body: notification.body,
        sound: notification.sound || 'default',
        icon: notification.icon || 'ic_notification',
      },
      data: {
        ...data,
        click_action: 'FLUTTER_NOTIFICATION_CLICK',
      }
    });

    return new Promise((resolve, reject) => {
      const options = {
        hostname: 'fcm.googleapis.com',
        port: 443,
        path: '/fcm/send',
        method: 'POST',
        headers: {
          'Authorization': `key=${this.serverKey}`,
          'Content-Type': 'application/json',
          'Content-Length': Buffer.byteLength(payload)
        }
      };

      const req = https.request(options, (res) => {
        let data = '';
        
        res.on('data', (chunk) => {
          data += chunk;
        });
        
        res.on('end', () => {
          try {
            const response = JSON.parse(data);
            resolve({
              success: res.statusCode === 200,
              messageId: response.message_id,
              response: response
            });
          } catch (error) {
            reject({
              success: false,
              error: 'Invalid JSON response',
              rawResponse: data
            });
          }
        });
      });

      req.on('error', (error) => {
        reject({
          success: false,
          error: error.message
        });
      });

      req.write(payload);
      req.end();
    });
  }
}

module.exports = FCMService;

// Usage example:
/*
const FCMService = require('./fcm-service');

const fcm = new FCMService('YOUR_FCM_SERVER_KEY_HERE');

// Send to single device
fcm.sendNotification(
  'ePaxeXHZRoGyUnWbMXU2...', // device token
  {
    title: 'Hello from Go-Mailer!',
    body: 'Your notification message here',
    sound: 'default'
  },
  {
    campaign_id: 'welcome_campaign',
    user_id: 'user123'
  }
).then(result => {
  console.log('✅ Notification sent:', result);
}).catch(error => {
  console.error('❌ Failed to send:', error);
});
*/
